rcl/src/rcl/graph.c

```diff
@@ -375,6 +377,81 @@ rcl_count_subscribers(
   return rcl_convert_rmw_ret_to_rcl_ret(rmw_ret);
 }
 
+
+typedef rmw_ret_t (* get_topic_info_func_t)(
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  rmw_topic_info_array_t * info_array);
+
+rcl_ret_t
+__rcl_get_info_by_topic(
+  const rcl_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  rmw_topic_info_array_t * info_array,
+  get_topic_info_func_t get_topic_info)
+{
+  if (!rcl_node_is_valid(node)) {
+    return RCL_RET_NODE_INVALID;  // error already set.
+  }
+  const rcl_node_options_t * node_options = rcl_node_get_options(node);
+  if (!node_options) {
+    return RCL_RET_NODE_INVALID;  // shouldn't happen, but error is already set if so
+  }
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(topic_name, RCL_RET_INVALID_ARGUMENT);
+  rmw_error_string_t error_string;
+  rmw_ret_t rmw_ret = rmw_topic_info_array_check_zero(info_array);
+  if (rmw_ret != RMW_RET_OK) {
+    error_string = rmw_get_error_string();
+    rmw_reset_error();
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "rmw_topic_info_array_t must be zero initialized: %s,\n"
+      "Use rmw_get_zero_initialized_topic_info_array",
+      error_string.str);
+    return rcl_convert_rmw_ret_to_rcl_ret(rmw_ret);
+  }
+  rmw_ret = get_topic_info(
+    rcl_node_get_rmw_handle(node),
+    allocator,
+    topic_name,
+    no_mangle,
+    info_array);
+  if (rmw_ret != RMW_RET_OK) {
+    error_string = rmw_get_error_string();
+    rmw_reset_error();
+    RCL_SET_ERROR_MSG(error_string.str);
```

Comment:
@mm318 I think you don't have to re-set the error message, though it makes me wonder why `rcl` and `rmw` share error code and message storage through `rcutils`. @wjwwood @jacobperron thoughts?

Reply:
You don't have to do it right now, but I would recommend doing it, in order to future proof yourself in the event that we change it so that rmw errors differ from rcl in the future.

Reply:
I'll leave it there then.